### PR TITLE
Enable HIP

### PR DESCRIPTION
**Pull request by masterleinad**

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

**Comment by dalg24** [comment]
Retest this please

**Comment by Rombur** [comment]
Retest this please

**Comment by Rombur** [comment]
Retest this please

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
I tried understanding why compiling in `Debug` mode fails but seem to be stuck. Basically, we never get to https://github.com/arborx/ArborX/blob/4b8fe6ff6fd48b020ccc26a67df0d9c2792408a7/src/details/ArborX_DetailsTreeTraversal.hpp#L348-L358 but try to return directly from the do-while loop in the tree traversal. Thus, the output array is never filled.

For now, I would just do the CI in `Release` mode to move forward here.

File: docker/Dockerfile.hipcc [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    cd benchmark && \
+    git clone https://github.com/google/googletest.git -b release-1.10.0 && \
+    mkdir build && cd build && \
+    cmake -D CMAKE_BUILD_TYPE=Release -D CMAKE_INSTALL_PREFIX=${BENCHMARK_DIR} .. && \
+    make -j${NPROCS} && make install && \
+    rm -rf ${SCRATCH_DIR}
+
+# Install Kokkos
+ARG KOKKOS_VERSION=0fdbdeda3d4d2cb00720627e46e3f2ecfb33e858
+ARG KOKKOS_OPTIONS="-DKokkos_ENABLE_HIP=ON -DKokkos_ENABLE_LIBDL=OFF -DKokkos_ENABLE_PROFILING=OFF -DKokkos_ARCH_VEGA906=ON"
```
Comment (dalg24) [inline docker/Dockerfile.hipcc:89]: I thought the profiling issue was resolved

File: docker/Dockerfile.hipcc [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    gpg --verify ${CMAKE_SHA256}.asc ${CMAKE_SHA256} && \
+    grep ${CMAKE_SCRIPT} ${CMAKE_SHA256} | sha256sum --check && \
+    mkdir -p ${CMAKE_DIR} && \
+    sh ${CMAKE_SCRIPT} --skip-license --prefix=${CMAKE_DIR} && \
+    rm ${CMAKE_SCRIPT}
+ENV PATH=${CMAKE_DIR}/bin:$PATH
+
+# Install Boost
+ENV BOOST_DIR=/opt/boost
+RUN BOOST_VERSION=1.72.0 && \
```
Comment (dalg24) [inline docker/Dockerfile.hipcc:48]: 1.73 is out

File: .jenkins [hunk 3]
```diff
@@ -314,8 +314,67 @@ pipeline {
...
+                        }
+                    }
+                    steps {
+                        sh 'ccache --zero-stats'
+                        sh 'rm -rf build && mkdir -p build'
+                        dir('build') {
+                            sh '''
+                                cmake \
+                                    -D CMAKE_INSTALL_PREFIX=$ARBORX_DIR \
+                                    -D CMAKE_BUILD_TYPE=Release \
```
Comment (dalg24) [inline .jenkins:335]: Did you have a chance to try Debug on Tulip?

File: docker/Dockerfile.hipcc (hunk 1 above)
Comment (masterleinad) [reply docker/Dockerfile.hipcc:89]: Yes, enabling profiling should work.

File: .jenkins (hunk 3 above)
Comment (masterleinad) [reply .jenkins:335]: Yes, same problem.

File: .jenkins [hunk 4]
```diff
@@ -314,8 +314,67 @@ pipeline {
...
+                        dockerfile {
+                            filename "Dockerfile.hipcc"
+                            dir "docker"
+                            additionalBuildArgs '--build-arg BASE=rocm/dev-ubuntu-18.04:3.5'
+                            args '-v /tmp/ccache.kokkos:/tmp/ccache --device=/dev/kfd --device=/dev/dri --security-opt seccomp=unconfined --group-add video'
+                            label 'rocm-docker && vega'
+                        }
+                    }
+                    steps {
+                        sh 'ccache --zero-stats'
```
Comment (dalg24) [inline .jenkins:329]: Remind me did we have issues with ccache?

Comment (masterleinad) [reply .jenkins:329]: Yes, we had (both in `Kokkos` and here). Do you want me to remove everything `ccache` related here?

**Comment by masterleinad** [comment]
The problem with `Debug` mode just seems to be `NDEBUG` and not the optimization flag.

**Comment by masterleinad** [comment]
This is puzzling me again. I checked locally that the last commit passes if I copy the `CMake` configuration for `ArborX` exactly. At this point, I would just drop the last commit again if that is good enough for now.

**Comment by masterleinad** [comment]
I squashed the last commits.

**Comment by masterleinad** [comment]
At least using `KOKKOS_EXPECTS` with (`NDEBUG`) seems to work again.

**Comment by masterleinad** [comment]
OK. It really doesn't work without `NDEBUG`. I'll remove the last commit.

**Comment by dalg24** [comment]
Did you mean to keep `KOKKOS_EXPECTS`?

**Comment by masterleinad** [comment]
> Did you mean to keep `KOKKOS_EXPECTS`?

Let me see what it takes to get this working with the newer commit.

**Comment by masterleinad** [comment]
OK. I am happy with this now. I dropped `KOKKOS_EXPECTS` but we still need `NDEBUG`.

**Pull request merged**